cupy/core/_routines_manipulation.pyx

```diff
@@ -559,18 +559,31 @@ cpdef ndarray concatenate_method(tup, int axis):
     if not have_same_types:
         dtype = functools.reduce(numpy.promote_types,
                                  set([a.dtype for a in arrays]))
-    return _concatenate(arrays, axis, tuple(shape), dtype)
+    if out is None:
+        out = ndarray(shape, dtype=dtype)
+    else:
+        if len(out.shape) != len(shape):
+            raise ValueError('Output array has wrong dimensionality')
+        if out.shape != shape:
+            raise ValueError('Output array is the wrong shape')
+        if out.dtype.kind != dtype.kind:
+            raise TypeError('Cannot cast scalar from dtype(\'{}\')'
+                            ' to dtype(\'{}\') according to the'
+                            ' rule \'same_kind\''.format(dtype, out.dtype))
+
+    return _concatenate(arrays, axis, tuple(shape), dtype, out)
```

Comment:
I think `tuple(shape)` should be done earlier, maybe just after the `for` loop above.

(I don't know whether/how `out.shape != shape` works for a `vector`. I guess it's converted to a `tuple` there implicitly.)